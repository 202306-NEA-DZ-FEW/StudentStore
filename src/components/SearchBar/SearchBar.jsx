import React, { useState, useEffect } from "react";
import { ImSearch } from "react-icons/im";
import { collection, query, onSnapshot } from "firebase/firestore";
import { db } from "@/util/firebase";

const SearchBar = () => {
    const [searchTerm, setSearchTerm] = useState("");
    const [products, setProducts] = useState([]);
    const [loading, setLoading] = useState(false);

    const handleSearch = async () => {
        setLoading(true);
        let q;
        if (searchTerm.trim() === "") {
            // If the search term is empty, do not perform a search.
            setProducts([]);
            setLoading(false);
            return;
        }
        // Use array-contains for case-insensitive search
        q = query(collection(db, "products"));

        onSnapshot(q, (snapshot) => {
            const productsArray = [];
            snapshot.forEach((doc) => {
                const data = doc.data();
                if (
                    data.name.toLowerCase().includes(searchTerm.toLowerCase())
                ) {
                    productsArray.push({ id: doc.id, ...data });
                }
            });
            setProducts(productsArray);
            setLoading(false);
        });
    };

    useEffect(() => {
        // Call the search function when searchTerm changes
        handleSearch();
    }, [searchTerm]);

    // Determine if products not found message should be shown
    const showNoProductsMessage =
        !loading && searchTerm.trim() !== "" && products.length === 0;
    const showResultsContainer = products.length > 0;

    return (
        <div className='relative mx-auto max-w-md'>
            <div className='input-wrapper'>
                <input
                    type='text'
                    placeholder='Search here...'
                    value={searchTerm}
                    onChange={(e) => setSearchTerm(e.target.value)}
                    className='w-full border border-gray-300 rounded-full py-2 px-4 pl-8 focus:outline-none'
                />
                <div className='absolute right-3 top-1/2 transform -translate-y-1/2 text-gray-500'>
                    <ImSearch />
                </div>
            </div>

            {showResultsContainer && (
                <div className='product-container bg-white bg-opacity-80 rounded-lg shadow-lg p-4 absolute left-0 right-0'>
                    <ul>
                        {products.map((product) => (
                            <li key={product.id}>
                                {product.name
                                    .split(new RegExp(`(${searchTerm})`, "gi"))
                                    .map((text, index) =>
                                        text.toLowerCase() ===
                                        searchTerm.toLowerCase() ? (
                                            <span
                                                key={index}
                                                className='text-blue-500'
                                            >
                                                {text}
                                            </span>
                                        ) : (
                                            text
                                        )
                                    )}
                            </li>
                        ))}
                    </ul>
                </div>
            )}

            {showNoProductsMessage && (
                <p className='text-white bg-gray-500 p-2 rounded'>
                    No products found.
                </p>
            )}
        </div>
    );
};

export default SearchBar;
