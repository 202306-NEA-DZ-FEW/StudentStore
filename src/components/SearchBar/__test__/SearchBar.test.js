import renderer from "react-test-renderer";
import SearchBar from "../SearchBar";

it("renders correctly", () => {
    const tree = renderer.create(<SearchBar />).toJSON();
    expect(tree).toMatchSnapshot();
});
