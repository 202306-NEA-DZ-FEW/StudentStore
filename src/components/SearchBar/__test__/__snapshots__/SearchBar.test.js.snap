// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="relative mx-auto max-w-sm"
>
  <div
    className="input-wrapper"
  >
    <input
      className="w-full py-1 px-4 border-[#585785] border-2 dark:bg-white rounded-full italic  font-sm  focus:outline-none"
      onChange={[Function]}
      placeholder="Search here..."
      type="text"
      value=""
    />
    <div
      className="absolute right-3 top-1/2 transform -translate-y-1/2 text-gray-500"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        version="1.1"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15.504 13.616l-3.79-3.223c-0.392-0.353-0.811-0.514-1.149-0.499 0.895-1.048 1.435-2.407 1.435-3.893 0-3.314-2.686-6-6-6s-6 2.686-6 6 2.686 6 6 6c1.486 0 2.845-0.54 3.893-1.435-0.016 0.338 0.146 0.757 0.499 1.149l3.223 3.79c0.552 0.613 1.453 0.665 2.003 0.115s0.498-1.452-0.115-2.003zM6 10c-2.209 0-4-1.791-4-4s1.791-4 4-4 4 1.791 4 4-1.791 4-4 4z"
        />
      </svg>
    </div>
  </div>
</div>
`;
